lib/scholar/linear/bayesian_ridge_regression.ex

```diff
@@ -0,0 +1,480 @@
+    scaled_sigma = Nx.dot(vh, [0], vh / Nx.new_axis(eigenvals + lambda / alpha, -1), [0])
+    sigma = scaled_sigma / alpha
+    {coef, intercept, alpha, lambda, iter, has_converged, scores, sigma}
+  end
+
+  defnp update_coef(
+          x,
+          y,
+          n_samples,
+          n_features,
+          xt_y,
+          u,
+          vh,
+          eigenvals,
+          alpha,
+          lambda
+        ) do
+    scaled_eigens = eigenvals + lambda / alpha
+
+    coef =
+      if n_samples > n_features do
+        regularization = vh / Nx.new_axis(scaled_eigens, -1)
+        reg_transpose = Nx.dot(regularization, xt_y)
+        Nx.dot(vh, [0], reg_transpose, [0])
+      else
+        regularization = u / scaled_eigens
+        reg_transpose = Nx.dot(regularization, Nx.dot(u, [0], y, [0]))
+        Nx.dot(x, [0], reg_transpose, [0])
+      end
+
+    error = y - Nx.dot(x, coef)
+    squared_error = error ** 2
+    rmse = Nx.sum(squared_error)
+
+    {coef, rmse}
+  end
+
+  defnp log_marginal_likelihood(
+          coef,
+          rmse,
+          n_samples,
+          n_features,
+          eigenvals,
+          alpha,
+          lambda,
+          alpha_1,
+          alpha_2,
+          lambda_1,
+          lambda_2
+        ) do
+    logdet_sigma =
+      if n_samples > n_features do
+        -1 * Nx.sum(Nx.log(lambda + alpha * eigenvals))
+      else
+        broad_lambda = Nx.broadcast(lambda, {n_samples})
+        -1 * Nx.sum(Nx.log(broad_lambda + alpha * eigenvals))
+      end
+
+    score_lambda = lambda_1 * Nx.log(lambda) - lambda_2 * lambda
+    score_alpha = alpha_1 * Nx.log(alpha) - alpha_2 * alpha
+
+    score_parameters =
+      n_features * Nx.log(lambda) + n_samples * Nx.log(alpha) - alpha * rmse -
+        lambda * Nx.sum(coef ** 2)
+
+    score =
+      0.5 * (score_parameters + logdet_sigma - n_samples * Nx.log(2 * Nx.Constants.pi()))
+
+    score_alpha + score_lambda + score
+  end
+
+  @doc """
+  Makes predictions with the given `model` on input `x`.
+  ## Examples
+
+      iex> x = Nx.tensor([[1], [2], [6], [8], [10]])
+      iex> y = Nx.tensor([1, 2, 6, 8, 10])
+      iex> model = Scholar.Linear.BayesianRidgeRegression.fit(x, y)
+      iex> Scholar.Linear.BayesianRidgeRegression.predict(model, Nx.tensor([[1], [3], [4]]))
+      Nx.tensor(
+        [1.02969491481781, 3.0161972045898438, 4.009448528289795]  
+      )  
+  """
+  deftransform predict(%__MODULE__{coefficients: coeff, intercept: intercept} = _model, x) do
+    predict_n(coeff, intercept, x)
+  end
+
+  defnp predict_n(coeff, intercept, x), do: Nx.dot(x, [-1], coeff, [-1]) + intercept
+
+  # Implements sample weighting by rescaling inputs and
+  # targets by sqrt(sample_weight).
+  defnp rescale(x, y, sample_weights) do
+    case Nx.shape(sample_weights) do
+      {} = scalar ->
+        scalar = Nx.sqrt(scalar)
+        {scalar * x, scalar * y}
+
+      _ ->
+        scale = sample_weights |> Nx.sqrt() |> Nx.make_diagonal()
```

Comment:
This can be simplified by doing
```suggestion
scale = sample_weights |> Nx.sqrt() |> Nx.new_axis(1)
```
and using `*` instead of `Nx.dot`.

Reply:
I do not believe this change is possible.
Please note that `a` is a `{n_samples}` sized vector, `b` is an `{n_samples, n_features}` matrix and `sample_weights` is an `{n_samples}` sized vector.

`sample_weights |> Nx.sqrt() |> Nx.new_axis(1)` yields an `{n_samples, 1}` sized matrix, that cannot be multiplied with `a`.

As far as I can tell, `sample_weights |> Nx.sqrt() |> Nx.make_diagonal()` yields the only matrix that is dottable with both `a` and `b`.

An alternative would be keeping 2 `scale` tensors, one for `a` and one for `b`. I personally do not like this option as it would unbalance the function. You would have a different tensor for each piece of data, and the operations would look different than the other branch of the `case` statement.

Please, let me know if you think of a better way.

Reply:
You are correct, but we should be able to use `*` instead of `Nx.dot`. This should work:
```
defnp rescale(x, y, sample_weights) do
  factor = Nx.sqrt(sample_weights)
  x_scaled = case Nx.shape(factor) do
    {} -> factor * x
    _ -> Nx.new_axis(factor, 1) * x
  end
  y_scaled = factor * y
  {x_scaled, y_scaled}
end
```

Reply:
This works and I find it's a pretty clean solution. Thank you for your comments.